File: client/src/main/scala/stasis/client/ops/backup/Backup.scala
package stasis.client.ops.backup

import akka.actor.typed.{ActorSystem, SpawnProtocol}
import akka.stream._
import akka.stream.scaladsl.{Sink, Source}
import akka.{Done, NotUsed}
import stasis.client.collection.rules.Rule
import stasis.client.encryption.secrets.DeviceSecret
import stasis.client.model.DatasetMetadata
import stasis.client.ops.ParallelismConfig
import stasis.client.ops.backup.stages._
import stasis.client.ops.exceptions.{EntityProcessingFailure, OperationStopped}
import stasis.client.tracking.BackupTracker
import stasis.shared.model.datasets.{DatasetDefinition, DatasetEntry}
import stasis.shared.ops.Operation

import java.nio.file.Path
import scala.concurrent.{ExecutionContext, Future}
import scala.util.{Failure, Success}

class Backup(
  descriptor: Backup.Descriptor
)(implicit system: ActorSystem[SpawnProtocol.Command], parallelism: ParallelismConfig, providers: Providers)
    extends Operation { parent =>
  import Backup._

  private implicit val mat: Materializer = SystemMaterializer(system).materializer

  private val supervision: Supervision.Decider = {
    case e: EntityProcessingFailure =>
      system.log.error(
        "Backup stream encountered failure while processing entity [{}]: [{} - {}]; resuming",
        e.entity,
        e.cause.getClass.getSimpleName,
        e.cause.getMessage
      )
      providers.track.failureEncountered(e.entity, failure = e.cause)
      Supervision.Resume

    case e =>
      system.log.error("Backup stream encountered failure: [{} - {}]; resuming", e.getClass.getSimpleName, e.getMessage)
      providers.track.failureEncountered(failure = e)
      Supervision.Resume
  }

  private implicit val ec: ExecutionContext = system.executionContext

  override implicit val id: Operation.Id = Operation.generateId()

  override val `type`: Operation.Type = Operation.Type.Backup

  private implicit val killSwitch: SharedKillSwitch = KillSwitches.shared("backup-kill-switch")

  private val stream: Source[Done, NotUsed] =
    stages.entityDiscovery
      .via(stages.entityCollection)
      .via(stages.entityProcessing)
      .via(stages.metadataCollection)
      .via(stages.metadataPush)
      .via(killSwitch.flow)
      .withAttributes(ActorAttributes.supervisionStrategy(supervision))

  override def start(): Future[Done] =
    stream
      .runWith(Sink.ignore)
      .trackWith(providers.track)

  override def stop(): Unit =
    killSwitch.abort(OperationStopped(s"Operation [${id.toString}] stopped by user"))

  private object stages
      extends EntityDiscovery
      with EntityCollection
      with EntityProcessing
      with MetadataCollection
      with MetadataPush {
    override protected lazy val collector: EntityDiscovery.Collector = parent.descriptor.collector.asDiscoveryCollector
    override protected lazy val targetDataset: DatasetDefinition = parent.descriptor.targetDataset
    override protected lazy val latestEntry: Option[DatasetEntry] = parent.descriptor.latestEntry
    override protected lazy val latestMetadata: Option[DatasetMetadata] = parent.descriptor.latestMetadata
    override protected lazy val deviceSecret: DeviceSecret = parent.descriptor.deviceSecret
    override protected lazy val providers: Providers = parent.providers
    override protected lazy val parallelism: ParallelismConfig = parent.parallelism
    override protected lazy val maxChunkSize: Int = parent.descriptor.limits.maxChunkSize
    override protected lazy val maxPartSize: Long = parent.descriptor.limits.maxPartSize
    override implicit protected lazy val mat: Materializer = parent.mat
    override implicit protected lazy val ec: ExecutionContext = parent.system.executionContext
  }
}

object Backup {
  def apply(
    descriptor: Backup.Descriptor
  )(implicit system: ActorSystem[SpawnProtocol.Command], parallelism: ParallelismConfig, providers: Providers): Backup =
    new Backup(descriptor)

  final case class Descriptor(
    targetDataset: DatasetDefinition,
    latestEntry: Option[DatasetEntry],
    latestMetadata: Option[DatasetMetadata],
    deviceSecret: DeviceSecret,
    collector: Descriptor.Collector,
    limits: Limits
  )

  object Descriptor {
    sealed trait Collector {
      def asDiscoveryCollector: EntityDiscovery.Collector
    }
    object Collector {
      final case class WithRules(rules: Seq[Rule]) extends Collector {
        override def asDiscoveryCollector: EntityDiscovery.Collector =
          EntityDiscovery.Collector.WithRules(rules)
      }

      final case class WithEntities(entities: Seq[Path]) extends Collector {
        override def asDiscoveryCollector: EntityDiscovery.Collector =
          EntityDiscovery.Collector.WithEntities(entities)
      }
    }

    def apply(
      definition: DatasetDefinition.Id,
      collector: Descriptor.Collector,
      deviceSecret: DeviceSecret,
      limits: Limits
    )(implicit ec: ExecutionContext, providers: Providers): Future[Descriptor] =
      for {
        targetDataset <- providers.clients.api.datasetDefinition(definition = definition)
        latestEntry <- providers.clients.api.latestEntry(definition = definition, until = None)
        latestMetadata <- latestEntry match {
          case Some(latestEntry) => providers.clients.api.datasetMetadata(entry = latestEntry).map(Some.apply)
          case None              => Future.successful(None)
        }
      } yield {
        Descriptor(
          targetDataset = targetDataset,
          latestEntry = latestEntry,
          latestMetadata = latestMetadata,
          deviceSecret = deviceSecret,
          collector = collector,
          limits = limits
        )
      }
  }

  final case class Limits(
    maxChunkSize: Int,
    maxPartSize: Long
  )

  implicit class TrackedOperation(operation: Future[Done]) {
    def trackWith(track: BackupTracker)(implicit id: Operation.Id, ec: ExecutionContext): Future[Done] = {
      operation.onComplete {
        case Success(_) | Failure(_: OperationStopped) =>
          track.completed()

        case Failure(e: EntityProcessingFailure) =>
          track.failureEncountered(entity = e.entity, failure = e.cause)
          track.completed()

        case Failure(e) =>
          track.failureEncountered(failure = e)
          track.completed()
      }

      operation
    }
  }
}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/ops/backup/Backup.kt
package stasis.client_android.lib.ops.backup

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.plus
import stasis.client_android.lib.collection.rules.Rule
import stasis.client_android.lib.encryption.secrets.DeviceSecret
import stasis.client_android.lib.model.DatasetMetadata
import stasis.client_android.lib.model.server.datasets.DatasetDefinition
import stasis.client_android.lib.model.server.datasets.DatasetDefinitionId
import stasis.client_android.lib.model.server.datasets.DatasetEntry
import stasis.client_android.lib.ops.Operation
import stasis.client_android.lib.ops.OperationId
import stasis.client_android.lib.ops.backup.stages.EntityCollection
import stasis.client_android.lib.ops.backup.stages.EntityDiscovery
import stasis.client_android.lib.ops.backup.stages.EntityProcessing
import stasis.client_android.lib.ops.backup.stages.MetadataCollection
import stasis.client_android.lib.ops.backup.stages.MetadataPush
import stasis.client_android.lib.ops.exceptions.EntityProcessingFailure
import stasis.client_android.lib.tracking.state.BackupState
import stasis.client_android.lib.utils.Try
import stasis.client_android.lib.utils.Try.Companion.flatMap
import stasis.client_android.lib.utils.Try.Companion.map
import stasis.client_android.lib.utils.Try.Success
import java.nio.file.Path
import java.util.concurrent.atomic.AtomicReference

class Backup(
    private val descriptor: Descriptor,
    private val providers: Providers
) : Operation {
    override val id: OperationId = descriptor.collector.existingState()?.operation ?: Operation.generateId()

    override fun type(): Operation.Type = Operation.Type.Backup

    override fun start(withScope: CoroutineScope, f: (Throwable?) -> Unit) {
        require(jobRef.get() == null) { "Backup [$id] already started" }

        val supervisor = SupervisorJob()

        val job = (withScope + supervisor).async {
            try {
                when (descriptor.collector) {
                    is Descriptor.Collector.WithState -> Unit // do nothing
                    else -> providers.track.started(operation = id, definition = descriptor.targetDataset.id)
                }

                stages.entityDiscovery(id)
                    .let { flow -> stages.entityCollection(id, flow) }
                    .let { flow -> stages.entityProcessing(id, flow) }
                    .let { flow -> stages.metadataCollection(id, flow, descriptor.collector.existingState()) }
                    .let { flow -> stages.metadataPush(id, flow) }
                    .collect()

                providers.track.completed(id)
            } catch (e: EntityProcessingFailure) {
                providers.track.failureEncountered(id, e.entity, e.cause)
                throw e
            } catch (e: Throwable) {
                providers.track.failureEncountered(id, e)
                throw e
            }
        }

        job.invokeOnCompletion { e ->
            supervisor.cancel()
            f(e)
        }

        jobRef.set(job)
    }

    override fun stop() {
        val job = jobRef.get()
        require(job != null) { "Backup [$id] not started" }

        job.cancel(cause = CancellationException("Cancelled by user"))
    }

    private val jobRef: AtomicReference<Job> = AtomicReference()

    private val stages =
        object : EntityDiscovery, EntityCollection, EntityProcessing, MetadataCollection, MetadataPush {
            override val collector: EntityDiscovery.Collector = descriptor.collector.asDiscoveryCollector()
            override val targetDataset: DatasetDefinition = descriptor.targetDataset
            override val latestEntry: DatasetEntry? = descriptor.latestEntry
            override val latestMetadata: DatasetMetadata? = descriptor.latestMetadata
            override val deviceSecret: DeviceSecret = descriptor.deviceSecret
            override val providers: Providers = this@Backup.providers
            override val maxPartSize: Long = descriptor.limits.maxPartSize
        }

    data class Descriptor(
        val targetDataset: DatasetDefinition,
        val latestEntry: DatasetEntry?,
        val latestMetadata: DatasetMetadata?,
        val deviceSecret: DeviceSecret,
        val collector: Collector,
        val limits: Limits
    ) {
        companion object {
            suspend operator fun invoke(
                definition: DatasetDefinitionId,
                collector: Collector,
                deviceSecret: DeviceSecret,
                limits: Limits,
                providers: Providers
            ): Try<Descriptor> =
                providers.clients.api.datasetDefinition(definition = definition)
                    .flatMap { targetDataset ->
                        providers.clients.api.latestEntry(definition = definition, until = null)
                            .flatMap { latestEntry ->
                                val metadata = latestEntry?.let {
                                    providers.clients.api.datasetMetadata(entry = latestEntry)
                                } ?: Success<DatasetMetadata?>(null)

                                metadata.map { latestMetadata ->
                                    Descriptor(
                                        targetDataset = targetDataset,
                                        latestEntry = latestEntry,
                                        latestMetadata = latestMetadata,
                                        deviceSecret = deviceSecret,
                                        collector = collector,
                                        limits = limits
                                    )
                                }
                            }
                    }
        }

        sealed class Collector {
            abstract fun asDiscoveryCollector(): EntityDiscovery.Collector
            abstract fun existingState(): BackupState?

            data class WithRules(val rules: List<Rule>) : Collector() {
                override fun asDiscoveryCollector(): EntityDiscovery.Collector =
                    EntityDiscovery.Collector.WithRules(rules)

                override fun existingState(): BackupState? = null
            }

            data class WithEntities(val entities: List<Path>) : Collector() {
                override fun asDiscoveryCollector(): EntityDiscovery.Collector =
                    EntityDiscovery.Collector.WithEntities(entities)

                override fun existingState(): BackupState? = null
            }

            data class WithState(val state: BackupState) : Collector() {
                override fun asDiscoveryCollector(): EntityDiscovery.Collector =
                    EntityDiscovery.Collector.WithState(state)

                override fun existingState(): BackupState = state
            }
        }

        data class Limits(
            val maxPartSize: Long
        )
    }
}
